### ISSUE TITLE
Unexpected end of file

### ISSUE THREAD

**Issue by rugk** [AtomLinter/linter-htmlhint#50]
This error was showed after I created a HTML file with markdown-preview.

```
SyntaxError: Unexpected end of input
  at Object.parse (native)
  at D:\Users\%username%\.atom\packages\linter-htmlhint\lib\init.coffee:41:32
  at process._tickCallback (node.js:365:9)
```

Atom 1.1.0
Windows 7

**Comment by cuplizian** [comment]
Same here, but with PHP inside HTML file.
Atom 1.2.0 on OS X El Capitan 10.11.1
linter-htmlhint 0.2.1

**Comment by johnwebbcole** [comment]
Could you post a short example.  I can't get any samples to throw the same error.

**Comment by abraham** [comment]
I was just digging around and the response from `/Users/abraham/.atom/packages/linter-htmlhint/node_modules/htmlhint/bin/htmlhint` contains unescaped `\n` instead of escaped `\\n` sometimes. This was causing [L41](https://github.com/AtomLinter/linter-htmlhint/blob/master/lib/init.coffee#L41) to blow up.

**Comment by abraham** [comment]
I have a snippet of code that I can't share but I'll try and get a sharable repo snippet later today.

**Comment by rugk** [comment]
> Could you post a short example.

In my test it works with every file produced by markdown-preview.

**Comment by abraham** [comment]
This triggers the error for me consistently. It looks like it's not actually an issue with not escaping newlines but instead is an issue with the response getting cropped somewhere.

_test.html.erb_
```
<% if @people.length == 0 %>
<% else %>
  <% @people.each do |p| %>
    <% p.contacts.each do |c| %>
      <tr class="<%= cycle("white", "shaded") %>"></tr>
    <% end %>
  <% end %>
<% end %>

```

This is what `output` is at [L41](https://github.com/AtomLinter/linter-htmlhint/blob/master/lib/init.coffee#L41) and you can see the end of it does not close up the JSON properly.


```
"[{"file":"/Users/abraham/dev/sandbox/test.html.erb","messages":[{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"<% if @people.length == 0 %>","line":1,"col":1,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"<% if @people.length == 0 %>","line":1,"col":28,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"<% else %>","line":2,"col":1,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"<% else %>","line":2,"col":10,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"  <% @people.each do |p| %>","line":3,"col":3,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"  <% @people.each do |p| %>","line":3,"col":27,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"    <% p.contacts.each do |c| %>","line":4,"col":5,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"    <% p.contacts.each do |c| %>","line":4,"col":32,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\"></tr>","line":5,"col":7,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\"></tr>","line":5,"col":18,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\"></tr>","line":5,"col":48,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\"></tr>","line":5,"col":50,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Doctype must be declared first.","raw":"<% if @people.length == 0 %>\n<% else %>\n  <% @people.each do |p| %>\n    <% p.contacts.each do |c| %>\n      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\">","evidence":"<% if @people.length == 0 %>","line":1,"col":1,"rule":{"id":"doctype-first","description":"Doctype must be declared first.","link":"https://github.com/yaniswang/HTMLHint/wiki/doctype-first"}},{"type":"error","message":"Tag must be paired, no start tag: [ </tr> ]","raw":"</tr>","evidence":"      <tr class=\"<%= cycle(\"white\", \"shaded\") %>\"></tr>","line":5,"col":51,"rule":{"id":"tag-pair","description":"Tag must be paired.","link":"https://github.com/yaniswang/HTMLHint/wiki/tag-pair"}},{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"\n    <% end %>\n  <% end %>\n<% end %>\n","evidence":"    <% end %>","line":6,"col":5,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"\n    <% end %>\n  <% end %>\n<% end %>\n","evidence":"    <% end %>","line":6,"col":13,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"\n    <% end %>\n  <% end %>\n<% end %>\n","evidence":"  <% end %>","line":7,"col":3,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"\n    <% end %>\n  <% end %>\n<% end %>\n","evidence":"  <% end %>","line":7,"col":11,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ < ].","raw":"\n    <% end %>\n  <% end %>\n<% end %>\n","evidence":"<% end %>","line":8,"col":1,"rule":{"id":"spec-char-escape","description":"Special characters must be escaped.","link":"https://github.com/yaniswang/HTMLHint/wiki/spec-char-escape"}},{"type":"error","message":"Special characters must be escaped : [ > ].","raw":"\n    <% end %>\n  <% end %>\n<% end %>\n","evidence":"<% end %>","line":8,"col":9,"rule":{"id":"spec-char-escape","description":"Spe"
```

Running `/Users/abraham/.atom/packages/linter-htmlhint/node_modules/htmlhint/bin/htmlhint test.html.erb --format json | jq` in the terminal results in a complete JSON result so it doesn't appear to be an issue with HTMLHint.

**Comment by johnwebbcole** [comment]
Yep, that threw an error for me too.  Thx!  I'll look at it.

**Comment by abraham** [comment]
Digging around a little more it seems like the `BufferedNodeProcess` `stdout` callback is getting the encoded JSON already truncated in [helpers.coffee](https://github.com/AtomLinter/atom-linter/blob/master/lib/helpers.coffee).

**Comment by rugk** [comment]
FYI this also happens when I create a completely empty file.
I created a `index.htm` file in a folder using the Atom treeview and directly after the file was opened these two errors were thrown:
![htmlempty](https://cloud.githubusercontent.com/assets/11966684/11349073/8a11a750-922a-11e5-8ee1-a328b1b350c0.PNG)

**Comment by admosity** [comment]
Should've looked this up first hah. I believe @abraham is right. I think `process.exit` is being triggered before the process has finished writing to `stdout`. `BufferedNodeProcess` listens for the `exit` event from the process. Despite there being more to capture, it stops it there and concats the input.

I've created a pull request in htmlhint to make sure stdout has been drained before exiting, but I'm about to close that because he moved over to streams for the formatters and my solution isn't going to work in that version [need more coffee](https://github.com/yaniswang/HTMLHint/pull/101).

So here's a gist instead of my fix of the current htmlhint version that I installed from npm (version 0.9.12). 
https://gist.github.com/admosity/f34bb48ed859571b6533

Specifically for the json formatter it checks if the write completed. 

Changes I made to the htmlhint executable:
![image](https://cloud.githubusercontent.com/assets/4655972/11574267/8e9b40e6-99d8-11e5-81f6-b060438cddfe.png)

![image](https://cloud.githubusercontent.com/assets/4655972/11574277/a2b9872c-99d8-11e5-847b-521272091d65.png)

**Comment by Arcanemagus** [comment]
Wow, that's quite a bug. Thanks for tracking this down @abraham and @admosity, I'm marking this as closed since it isn't a bug with `linter-htmlhint`.

**Comment by admosity** [comment]
No problem! I've updated the gist to more reliable code. It appears that `BufferedProcess`'s child process spawn is getting truncated data. I can't seem to replicate the behavior on node 0.12.7 and I'm able to retrieve all the data despite the early exit. 

Also looks like even after a child process's exit event is fired, its stdout can remain open. 
https://nodejs.org/api/child_process.html#child_process_event_exit
Which `BufferedProcess` accounts for by checking:
https://github.com/atom/atom/blob/master/src/buffered-process.coffee#L200,L232
That all streams are closed before firing the exit callback.
